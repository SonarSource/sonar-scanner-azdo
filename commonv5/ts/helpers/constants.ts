export const PROP_NAMES = {
  HOST_URL: "sonar.host.url",
  TOKEN: "sonar.token",
  LOGIN: "sonar.login",
  PASSSWORD: "sonar.password",
  ORG: "sonar.organization",
  PROJECTKEY: "sonar.projectKey",
  PROJECTNAME: "sonar.projectName",
  PROJECTVERSION: "sonar.projectVersion",
  PROJECTSOURCES: "sonar.sources",
  PROJECTSETTINGS: "project.settings",
};

export enum AzureProvider {
  TfsGit = "TfsGit",
  Svn = "Svn",
  Git = "Git",
  GitHub = "GitHub",
  GitHubEnterprise = "GitHubEnterprise",
  Bitbucket = "Bitbucket",
}

export const DEFAULT_BRANCH_REF = "refs/heads/master";

export enum AzureBuildVariables {
  BuildRepositoryName = "Build.Repository.Name",
}

export enum TaskVariables {
  SonarQubeServerVersion = "SONARQUBE_SERVER_VERSION",
  SonarQubeScannerMode = "SONARQUBE_SCANNER_MODE",
  SonarQubeScannerParams = "SONARQUBE_SCANNER_PARAMS",
  SonarQubeScannerReportTaskFile = "SONARQUBE_SCANNER_REPORTTASKFILE",
  SonarQubeEndpoint = "SONARQUBE_ENDPOINT",
  SonarQubeScannerMSBuildExe = "SONARQUBE_SCANNER_MSBUILD_EXE",
  SonarQubeScannerMSBuildDll = "SONARQUBE_SCANNER_MSBUILD_DLL",
  JavaHome = "JAVA_HOME",
}

export const TASK_MISSING_VARIABLE_ERROR_HINT = `Make sure you are not mixing tasks from different major versions. If you are using a multistage pipeline, make sure the tasks are in the same stage.`;
