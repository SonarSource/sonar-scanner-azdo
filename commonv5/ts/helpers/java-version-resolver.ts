import * as tl from "azure-pipelines-task-lib/task";

export default class JavaVersionResolver {
  private static javaHomeOriginalPath: string;
  private static isJavaNewVersionSet: boolean = false;

  public static setJavaHomeToIfAvailable(newJavaPath: string) {
    this.javaHomeOriginalPath = tl.getVariable("JAVA_HOME");
    const javaNewPath = tl.getVariable(newJavaPath);
    if (javaNewPath) {
      tl.debug(`${newJavaPath} path has been detected, switching to it for the analysis.`);
      tl.setVariable("JAVA_HOME", javaNewPath);
      this.isJavaNewVersionSet = true;
    }
  }

  public static revertJavaHomeToOriginal() {
    if (this.isJavaNewVersionSet) {
      tl.debug("Reverting JAVA_HOME to its initial path.");
      tl.setVariable("JAVA_HOME", this.javaHomeOriginalPath);
    }
  }
}
