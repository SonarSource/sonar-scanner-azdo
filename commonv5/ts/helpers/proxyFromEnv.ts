import { URL } from "url";

function formatHostname(hostname: string) {
  // canonicalize the hostname, so that 'oogle.com' won't match 'google.com'
  return hostname.replace(/^\.*/, ".").toLowerCase();
}

function parseNoProxyZone(zone: string) {
  zone = zone.trim().toLowerCase();

  const zoneParts = zone.split(":", 2);
  const zoneHost = formatHostname(zoneParts[0]);
  const zonePort = zoneParts[1];
  const hasPort = zone.indexOf(":") > -1;

  return { hostname: zoneHost, port: zonePort, hasPort };
}

function uriInNoProxy(url: URL, noProxy: string) {
  const port = url.port || (url.protocol === "https:" ? "443" : "80");
  const hostname = formatHostname(url.hostname);
  const noProxyList = noProxy.split(",");

  // iterate through the noProxyList until it finds a match.
  return noProxyList.map(parseNoProxyZone).some((noProxyZone) => {
    const isMatchedAt = hostname.indexOf(noProxyZone.hostname);
    const hostnameMatched =
      isMatchedAt > -1 && isMatchedAt === hostname.length - noProxyZone.hostname.length;

    if (noProxyZone.hasPort) {
      return port === noProxyZone.port && hostnameMatched;
    }

    return hostnameMatched;
  });
}

export function getProxyFromURI(url: URL) {
  // Decide the proper request proxy to use based on the request URI object and the
  // environmental constiables (NO_PROXY, HTTP_PROXY, etc.)
  // respect NO_PROXY environment constiables (see: https://lynx.invisible-island.net/lynx2.8.7/breakout/lynx_help/keystrokes/environments.html)

  const noProxy = process.env.NO_PROXY || process.env.no_proxy || "";

  // if the noProxy is a wildcard then return null

  if (noProxy === "*") {
    return null;
  }

  // if the noProxy is not empty and the uri is found return null

  if (noProxy !== "" && uriInNoProxy(url, noProxy)) {
    return null;
  }

  // Check for HTTP or HTTPS Proxy in environment Else default to null

  if (url.protocol === "http:") {
    return process.env.HTTP_PROXY || process.env.http_proxy || null;
  }

  if (url.protocol === "https:") {
    return (
      process.env.HTTPS_PROXY ||
      process.env.https_proxy ||
      process.env.HTTP_PROXY ||
      process.env.http_proxy ||
      null
    );
  }

  // if none of that works, return null
  // (What uri protocol are you using then?)

  return null;
}
