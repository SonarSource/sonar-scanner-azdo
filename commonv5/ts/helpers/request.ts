import * as tl from "azure-pipelines-task-lib/task";
import fetch from "node-fetch";
import * as semver from "semver";
import Endpoint from "../sonarqube/Endpoint";

interface RequestData {
  [x: string]: any;
}

export async function get<T>(
  endpoint: Endpoint,
  path: string,
  isJson: boolean,
  query?: RequestData,
): Promise<T | string> {
  const hasQuery = query && Object.keys(query).length > 0;
  const fullUrl =
    endpoint.url + path + (hasQuery ? "?" + new URLSearchParams(query).toString() : "");
  tl.debug(
    `[SQ] API GET: '${path}' with full URL "${fullUrl}" and query "${JSON.stringify(query)}"`,
  );

  try {
    const response = await fetch(fullUrl, endpoint.toFetchOptions(fullUrl));
    if (isJson) {
      return await response.json();
    } else {
      return await response.text();
    }
  } catch (error) {
    if (error.response) {
      tl.debug(`[SQ] API GET '${path}' failed, status code was: ${error.response.status}`);
    } else {
      tl.debug(`[SQ] API GET '${path}' failed, error is ${error.message}`);
    }
    throw new Error(`[SQ] API GET '${path}' failed, error is ${error.message}`);
  }
}

export async function getServerVersion(endpoint: Endpoint): Promise<semver.SemVer> {
  const serverVersion = await get<string>(endpoint, "/api/server/version", false);
  tl.debug(`[SQ] Server version: ${serverVersion}`);
  return semver.coerce(serverVersion);
}
