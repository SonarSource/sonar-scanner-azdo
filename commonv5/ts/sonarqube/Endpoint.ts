import * as tl from "azure-pipelines-task-lib/task";
import { PROP_NAMES } from "../helpers/utils";

export enum EndpointType {
  SonarCloud = "SonarCloud",
  SonarQube = "SonarQube",
}

export interface EndpointData {
  url: string;
  token?: string;
  username?: string;
  password?: string;
  organization?: string;
}

export default class Endpoint {
  constructor(public type: EndpointType, private readonly data: EndpointData) {}

  public get auth() {
    if (!this.data.token && this.data.password && this.data.password.length > 0) {
      return { user: this.data.username, pass: this.data.password };
    }
    return { user: this.data.token || this.data.username };
  }

  public get organization() {
    return this.data.organization;
  }

  public get url() {
    return this.data.url;
  }

  public toJson() {
    return JSON.stringify({ type: this.type, data: this.data });
  }

  public toSonarProps() {
    return {
      [PROP_NAMES.HOST_URL]: this.data.url,
      [PROP_NAMES.LOGIN]: this.data.token || this.data.username,
      [PROP_NAMES.PASSSWORD]:
        this.data.password && this.data.password.length > 0 ? this.data.password : null,
      [PROP_NAMES.ORG]: this.data.organization,
    };
  }

  public static getEndpoint(id: string, type: EndpointType): Endpoint {
    const url = tl.getEndpointUrl(id, false);
    const token = tl.getEndpointAuthorizationParameter(
      id,
      "apitoken",
      type !== EndpointType.SonarCloud
    );
    const username = tl.getEndpointAuthorizationParameter(id, "username", true);
    const password = tl.getEndpointAuthorizationParameter(id, "password", true);
    const organization = tl.getInput("organization", type === EndpointType.SonarCloud);
    return new Endpoint(type, { url, token, username, password, organization });
  }
}
