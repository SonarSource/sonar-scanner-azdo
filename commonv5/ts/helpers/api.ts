import * as tl from "azure-pipelines-task-lib/task";
import { get } from "./request";
import Endpoint from "../sonarqube/Endpoint";
import { Metric, MetricsResponse, ProjectStatus } from "../sonarqube/types";

export async function fetchProjectStatus(
  endpoint: Endpoint,
  analysisId: string,
): Promise<ProjectStatus> {
  tl.debug(`Retrieve Analysis id '${analysisId}.'`);

  try {
    const { projectStatus } = (await get(endpoint, "/api/qualitygates/project_status", true, {
      analysisId,
    })) as { projectStatus: ProjectStatus };
    return projectStatus;
  } catch (error) {
    if (error?.message) {
      tl.error(`Error retrieving analysis: ${error.message}`);
    } else if (error) {
      tl.error(`Error retrieving analysis: ${JSON.stringify(error)}`);
    }
    throw new Error(`Could not fetch analysis for ID '${analysisId}'`);
  }
}

export async function fetchMetrics(
  endpoint: Endpoint,
  data: { f?: string; p?: number; ps?: number } = { f: "name", ps: 500 },
  prev?: MetricsResponse,
): Promise<Metric[]> {
  try {
    const response = (await get(endpoint, "/api/metrics/search", true, data)) as MetricsResponse;
    const { metrics, p, ps, total } = response;
    const result = prev ? prev.metrics.concat(metrics) : metrics;
    if (p * ps >= total) {
      return result;
    }
    return fetchMetrics(endpoint, { ...data, p: p + 1 }, { ...response, metrics: result });
  } catch (error) {
    if (error?.message) {
      tl.error(error.message);
    } else if (error) {
      tl.error(JSON.stringify(error));
    }

    throw new Error(`Could not fetch metrics`);
  }
}
