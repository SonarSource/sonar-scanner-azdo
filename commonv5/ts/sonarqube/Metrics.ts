import * as tl from "azure-pipelines-task-lib/task";
import Endpoint from "./Endpoint";
import { getJSON } from "../helpers/request";

interface IMetric {
  custom?: boolean;
  decimalScale?: number;
  description?: string;
  direction?: number;
  domain?: string;
  hidden?: boolean;
  key: string;
  name: string;
  qualitative?: boolean;
  type: string;
}

interface MetricsResponse {
  metrics: IMetric[];
  p: number;
  ps: number;
  total: number;
}

export default class Metrics {
  constructor(public metrics: IMetric[]) {}

  public getMetricByKey(key: string) {
    return this.metrics.find((metric) => metric.key === key);
  }

  public static getAllMetrics(endpoint: Endpoint): Promise<Metrics | undefined> {
    return inner().catch((err) => {
      tl.error(`[SQ] Could not fetch metrics`);
      if (err && err.message) {
        tl.error(err.message);
      } else if (err) {
        tl.error(JSON.stringify(err));
      }
      return undefined;
    });

    function inner(
      data: { f?: string; p?: number; ps?: number } = { f: "name", ps: 500 },
      prev?: MetricsResponse
    ): Promise<Metrics> {
      return getJSON(endpoint, "/api/metrics/search", data).then((r: MetricsResponse) => {
        const result = prev ? prev.metrics.concat(r.metrics) : r.metrics;
        if (r.p * r.ps >= r.total) {
          return new Metrics(result);
        }
        return inner({ ...data, p: r.p + 1 }, { ...r, metrics: result });
      });
    }
  }
}
