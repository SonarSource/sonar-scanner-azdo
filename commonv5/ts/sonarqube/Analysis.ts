import * as tl from "azure-pipelines-task-lib/task";
import Endpoint, { EndpointType } from "./Endpoint";
import Metrics from "./Metrics";
import { formatMeasure } from "../helpers/measures";
import { getJSON } from "../helpers/request";

interface IAnalysis {
  status: string;
  conditions: Condition[];
}

interface Condition {
  status: string;
  metricKey: string;
  actualValue?: string;
  comparator?: string;
  periodIndex?: number;
  errorThreshold?: string;
  warningThreshold?: string;
}

export default class Analysis {
  constructor(
    private readonly analysis: IAnalysis,
    private readonly endpointType: EndpointType,
    private readonly warnings: string[],
    private readonly dashboardUrl?: string,
    private readonly metrics?: Metrics,
    private readonly projectName?: string
  ) {}

  public get status() {
    return this.analysis.status.toUpperCase();
  }

  public get conditions() {
    return this.analysis.conditions;
  }

  public getFailedConditions() {
    return this.conditions.filter((condition) =>
      ["WARN", "ERROR"].includes(condition.status.toUpperCase())
    );
  }

  public getHtmlAnalysisReport() {
    tl.debug(`[SQ] Generate analysis report.'`);
    return [
      this.getQualityGateSection(),
      this.getQualityGateDetailSection(),
      this.getDashboardLink(),
      this.getWarnings(),
    ]
      .join(" \r\n")
      .trim();
  }

  private getQualityGateSection() {
    const qgStyle = `background-color: ${this.getQualityGateColor()};
      padding: 4px 12px;
      color: #fff;
      letter-spacing: 0.02em;
      line-height: 24px;
      font-weight: 600;
      font-size: 12px;
      margin-left: 15px;`;
    return `<div style="padding-top: 8px;">
      <span>${this.projectName ? this.projectName + " " : ""}Quality Gate</span>
      <span style="${qgStyle}">
        ${formatMeasure(this.status, "LEVEL")}
      </span>
    </div>`;
  }

  private getQualityGateDetailSection() {
    const failedConditions = this.getFailedConditions();
    if (!this.metrics || !["WARN", "ERROR"].includes(this.status) || failedConditions.length <= 0) {
      return "";
    }

    const rows = failedConditions.map((condition) => {
      const metric = this.metrics && this.metrics.getMetricByKey(condition.metricKey);
      if (!metric) {
        return null;
      }

      const threshold =
        condition.status === "WARN" ? condition.warningThreshold : condition.errorThreshold;
      return `<tr>
        <td><span style="padding-right:4px;">${metric.name}</span></td>
        <td style="text-align:center; color:#fff; background-color:${this.getQualityGateColor()};">
          <span style="padding:0px 4px; line-height:20px;">${formatMeasure(
            condition.actualValue,
            metric.type
          )}</span>
        </td>
        <td>
          <span style="padding-left:4px">${formatMeasure(condition.comparator, "COMPARATOR")}</span>
          <span style="padding-left:4px">${formatMeasure(threshold, metric.type)}</span>
        </td>
      </tr>`;
    });

    const tableStyle = `
      margin-top: 8px;
      border-top: 1px solid #eee;
      border-collapse: separate;
      border-spacing: 0 4px;
    `;
    return `<table border="0" style="${tableStyle}">
      <tbody>
        ${rows.join(" \r\n").trim()}
      </tbody>
    </table>\r\n\r\n`; // 2 carriage returns to prevent any malformed summary results
  }

  private getDashboardLink() {
    if (!this.dashboardUrl) {
      return "";
    }
    const linkText = `Detailed ${this.endpointType} report &gt;`;
    return `[${linkText}](${this.dashboardUrl})`;
  }

  public getWarnings() {
    if (this.warnings.some((w) => w.includes("Please update to at least Java 11"))) {
      return `<br><span>&#9888;</span><b>${this.warnings.find((w) =>
        w.includes("Please update to at least Java 11")
      )}</b>`;
    } else {
      return "";
    }
  }

  private getQualityGateColor() {
    switch (this.status) {
      case "OK":
        return "#00aa00";
      case "WARN":
        return "#ed7d20";
      case "ERROR":
        return "#d4333f";
      case "NONE":
        return "#b4b4b4";
      default:
        return "#b4b4b4";
    }
  }

  public static getAnalysis({
    analysisId,
    projectName,
    endpoint,
    metrics,
    dashboardUrl,
    warnings,
  }: {
    analysisId: string;
    dashboardUrl?: string;
    endpoint: Endpoint;
    projectName?: string;
    metrics?: Metrics;
    warnings: string[];
  }): Promise<Analysis> {
    tl.debug(`[SQ] Retrieve Analysis id '${analysisId}.'`);
    return getJSON(endpoint, "/api/qualitygates/project_status", { analysisId }).then(
      ({ projectStatus }: { projectStatus: IAnalysis }) =>
        new Analysis(projectStatus, endpoint.type, warnings, dashboardUrl, metrics, projectName),
      (err) => {
        if (err && err.message) {
          tl.error(`[SQ] Error retrieving analysis: ${err.message}`);
        } else if (err) {
          tl.error(`[SQ] Error retrieving analysis: ${JSON.stringify(err)}`);
        }
        throw new Error(`[SQ] Could not fetch analysis for ID '${analysisId}'`);
      }
    );
  }
}
