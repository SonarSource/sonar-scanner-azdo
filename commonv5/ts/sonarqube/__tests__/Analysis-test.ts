import Analysis from "../Analysis";
import { getJSON } from "../../helpers/request";
import Metrics from "../Metrics";
import Endpoint, { EndpointType } from "../Endpoint";

jest.mock("../../helpers/request", () => ({
  getJSON: jest.fn(() =>
    Promise.resolve({
      projectStatus: {
        status: "ERROR",
        conditions: [
          {
            status: "ERROR",
            metricKey: "bugs",
            comparator: "GT",
            errorThreshold: "0",
            actualValue: "1",
          },
        ],
      },
    })
  ),
}));

jest.mock("azure-pipelines-task-lib/task", () => ({
  debug: jest.fn(),
  error: jest.fn(),
}));

const METRICS = new Metrics([{ key: "bugs", name: "Bugs", type: "INT" }]);
const ENDPOINT = new Endpoint(EndpointType.SonarQube, { url: "https://endpoint.url" });
const GET_ANALYSIS_DATA = {
  analysisId: "analysisId",
  dashboardUrl: "https://dashboard.url",
  endpoint: ENDPOINT,
  metrics: METRICS,
  warnings: [],
};

beforeEach(() => {
  (getJSON as jest.Mock<any>).mockClear();
});

it("should generate an analysis status with error", async () => {
  const analysis = await Analysis.getAnalysis(GET_ANALYSIS_DATA);
  expect(getJSON).toHaveBeenCalledWith(ENDPOINT, "/api/qualitygates/project_status", {
    analysisId: "analysisId",
  });
  expect(analysis.status).toBe("ERROR");
  expect(analysis.getFailedConditions()).toHaveLength(1);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should generate a green analysis status", async () => {
  (getJSON as jest.Mock<any>).mockImplementationOnce(() =>
    Promise.resolve({ projectStatus: { status: "SUCCESS", conditions: [] } })
  );

  const analysis = await Analysis.getAnalysis(GET_ANALYSIS_DATA);
  expect(getJSON).toHaveBeenCalledWith(ENDPOINT, "/api/qualitygates/project_status", {
    analysisId: "analysisId",
  });
  expect(analysis.status).toBe("SUCCESS");
  expect(analysis.getFailedConditions()).toHaveLength(0);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should not fail when metrics are missing", async () => {
  const analysis = await Analysis.getAnalysis({
    ...GET_ANALYSIS_DATA,
    dashboardUrl: undefined,
    metrics: undefined,
  });
  expect(getJSON).toHaveBeenCalledWith(ENDPOINT, "/api/qualitygates/project_status", {
    analysisId: "analysisId",
  });
  expect(analysis.status).toBe("ERROR");
  expect(analysis.getFailedConditions()).toHaveLength(1);
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should display the project name", async () => {
  const analysis = await Analysis.getAnalysis({
    ...GET_ANALYSIS_DATA,
    projectName: "project_name",
  });
  expect(analysis.getHtmlAnalysisReport()).toMatchSnapshot();
});

it("should display Java 11 warning", async () => {
  const analysis = await Analysis.getAnalysis({
    ...GET_ANALYSIS_DATA,
    warnings: [
      "The version of Java (1.8.0_221) you have used to run this analysis is deprecated and we will stop accepting it from October 2020. Please update to at least Java 11.",
    ],
  });
  expect(analysis.getWarnings()).toStrictEqual(
    "<br><span>&#9888;</span><b>The version of Java (1.8.0_221) you have used to run this analysis is deprecated and we will stop accepting it from October 2020. Please update to at least Java 11.</b>"
  );
});
