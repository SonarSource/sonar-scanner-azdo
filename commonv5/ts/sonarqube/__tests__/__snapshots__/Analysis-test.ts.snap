// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the project name 1`] = `
"<div style="
  color: #3E4357;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
      <span>project_name Quality Gate</span>
      <span style="background-color: #D92D20;
  padding: 4px 8px;
  margin: 0px 4px;
  color: #fff;
  letter-spacing: -0.12px;
  line-height: normal;
  font-weight: 500;
  font-size: 12px;">
        Failed
      </span>
    </div> 
<div style="
  background: #EBECF0;
  height: 1px;
  margin: 16px 0px;"></div> 
<ul style="
  color: #3E4357;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
        <li style="list-style: none; margin-top: 13px;">
        <span>Bugs</span>
        <span style="background-color: #D92D20;
  padding: 4px 8px;
  margin: 0px 4px;
  color: #fff;
  letter-spacing: -0.12px;
  line-height: normal;
  font-weight: 500;
  font-size: 12px;">
          1</span>
        <span>(required &le; 0)</span>
      </li>
    </ul>

 
<div style="margin-top: 16px;">
        <a href="https://dashboard.url" style="
  color: #5D6CD0;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
        Detailed SonarQube report
        <span class="bowtie-icon bowtie-navigate-external" style="color: #5D6CD0; font-size: 20px; vertical-align: middle;"></span>
        </a>
      </div>"
`;

exports[`should generate a green analysis status 1`] = `
"<div style="
  color: #3E4357;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
      <span>Quality Gate</span>
      <span style="background-color: #b4b4b4;
  padding: 4px 8px;
  margin: 0px 4px;
  color: #fff;
  letter-spacing: -0.12px;
  line-height: normal;
  font-weight: 500;
  font-size: 12px;">
        SUCCESS
      </span>
    </div> 
<div style="
  background: #EBECF0;
  height: 1px;
  margin: 16px 0px;"></div> 
 
<div style="margin-top: 16px;">
        <a href="https://dashboard.url" style="
  color: #5D6CD0;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
        Detailed SonarQube report
        <span class="bowtie-icon bowtie-navigate-external" style="color: #5D6CD0; font-size: 20px; vertical-align: middle;"></span>
        </a>
      </div>"
`;

exports[`should generate an analysis status with error 1`] = `
"<div style="
  color: #3E4357;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
      <span>Quality Gate</span>
      <span style="background-color: #D92D20;
  padding: 4px 8px;
  margin: 0px 4px;
  color: #fff;
  letter-spacing: -0.12px;
  line-height: normal;
  font-weight: 500;
  font-size: 12px;">
        Failed
      </span>
    </div> 
<div style="
  background: #EBECF0;
  height: 1px;
  margin: 16px 0px;"></div> 
<ul style="
  color: #3E4357;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
        <li style="list-style: none; margin-top: 13px;">
        <span>Bugs</span>
        <span style="background-color: #D92D20;
  padding: 4px 8px;
  margin: 0px 4px;
  color: #fff;
  letter-spacing: -0.12px;
  line-height: normal;
  font-weight: 500;
  font-size: 12px;">
          1</span>
        <span>(required &le; 0)</span>
      </li>
    </ul>

 
<div style="margin-top: 16px;">
        <a href="https://dashboard.url" style="
  color: #5D6CD0;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
        Detailed SonarQube report
        <span class="bowtie-icon bowtie-navigate-external" style="color: #5D6CD0; font-size: 20px; vertical-align: middle;"></span>
        </a>
      </div>"
`;

exports[`should not fail when metrics are missing 1`] = `
"<div style="
  color: #3E4357;
  font-family: Inter;
  font-size: 14px;
  font-style: normal;
  font-weight: 500;
  line-height: normal;
  letter-spacing: -0.14px;">
      <span>Quality Gate</span>
      <span style="background-color: #D92D20;
  padding: 4px 8px;
  margin: 0px 4px;
  color: #fff;
  letter-spacing: -0.12px;
  line-height: normal;
  font-weight: 500;
  font-size: 12px;">
        Failed
      </span>
    </div> 
<div style="
  background: #EBECF0;
  height: 1px;
  margin: 16px 0px;"></div>"
`;
